File: app/src/main/java/me/devsaki/hentoid/fragments/pin/PinDialogFragment.java
package me.devsaki.hentoid.fragments.pin;

import android.content.Context;
import android.os.Bundle;
import android.os.Vibrator;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.fragment.app.DialogFragment;

import java.security.InvalidParameterException;

import me.devsaki.hentoid.R;
import me.devsaki.hentoid.util.ThemeHelper;

import static androidx.core.view.ViewCompat.requireViewById;

public abstract class PinDialogFragment extends DialogFragment {

    private final StringBuilder pinValue = new StringBuilder(4);
    private TextView headerText;
    private View placeholderImage1;
    private View placeholderImage2;
    private View placeholderImage3;
    private View placeholderImage4;

    protected abstract void onPinAccept(String pin);

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ThemeHelper.setStyle(requireActivity(), this, STYLE_NORMAL, R.style.Theme_Light_PinEntryDialog);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_pin_dialog, container, false);

        headerText = requireViewById(rootView, R.id.text_header);

        placeholderImage1 = requireViewById(rootView, R.id.image_placeholder_1);

        placeholderImage2 = requireViewById(rootView, R.id.image_placeholder_2);

        placeholderImage3 = requireViewById(rootView, R.id.image_placeholder_3);

        placeholderImage4 = requireViewById(rootView, R.id.image_placeholder_4);

        View button0 = requireViewById(rootView, R.id.button_0);
        button0.setOnClickListener(v -> onKeyClick("0"));

        View button1 = requireViewById(rootView, R.id.button_1);
        button1.setOnClickListener(v9 -> onKeyClick("1"));

        View button2 = requireViewById(rootView, R.id.button_2);
        button2.setOnClickListener(v8 -> onKeyClick("2"));

        View button3 = requireViewById(rootView, R.id.button_3);
        button3.setOnClickListener(v7 -> onKeyClick("3"));

        View button4 = requireViewById(rootView, R.id.button_4);
        button4.setOnClickListener(v6 -> onKeyClick("4"));

        View button5 = requireViewById(rootView, R.id.button_5);
        button5.setOnClickListener(v5 -> onKeyClick("5"));

        View button6 = requireViewById(rootView, R.id.button_6);
        button6.setOnClickListener(v4 -> onKeyClick("6"));

        View button7 = requireViewById(rootView, R.id.button_7);
        button7.setOnClickListener(v3 -> onKeyClick("7"));

        View button8 = requireViewById(rootView, R.id.button_8);
        button8.setOnClickListener(v2 -> onKeyClick("8"));

        View button9 = requireViewById(rootView, R.id.button_9);
        button9.setOnClickListener(v1 -> onKeyClick("9"));

        View buttonBackspace = requireViewById(rootView, R.id.button_backspace);
        buttonBackspace.setOnClickListener(v -> onBackspaceClick());

        return rootView;
    }

    @Override
    public void onStop() {
        super.onStop();
        getDialog().cancel();
    }

    void clearPin() {
        pinValue.setLength(0);
        placeholderImage1.setVisibility(View.INVISIBLE);
        placeholderImage2.setVisibility(View.INVISIBLE);
        placeholderImage3.setVisibility(View.INVISIBLE);
        placeholderImage4.setVisibility(View.INVISIBLE);
    }

    void setHeaderText(@StringRes int resId) {
        headerText.setText(resId);
    }

    void vibrate() {
        Vibrator vibrator = (Vibrator) requireContext().getSystemService(Context.VIBRATOR_SERVICE);
        if (vibrator != null) {
            vibrator.vibrate(300);
        }
    }

    private void onKeyClick(String s) {
        if (pinValue.length() == 4) return;

        pinValue.append(s);
        switch (pinValue.length()) {
            case 1:
                placeholderImage1.setVisibility(View.VISIBLE);
                break;
            case 2:
                placeholderImage2.setVisibility(View.VISIBLE);
                break;
            case 3:
                placeholderImage3.setVisibility(View.VISIBLE);
                break;
            case 4:
                placeholderImage4.setVisibility(View.VISIBLE);
                onPinAccept(pinValue.toString());
                break;
            default:
                throw new InvalidParameterException("Not implemented");
        }
    }

    private void onBackspaceClick() {
        if (pinValue.length() == 0) return;

        pinValue.setLength(pinValue.length() - 1);
        switch (pinValue.length()) {
            case 0:
                placeholderImage1.setVisibility(View.INVISIBLE);
                break;
            case 1:
                placeholderImage2.setVisibility(View.INVISIBLE);
                break;
            case 2:
                placeholderImage3.setVisibility(View.INVISIBLE);
                break;
            case 3:
                placeholderImage4.setVisibility(View.INVISIBLE);
                break;
            default:
                throw new InvalidParameterException("Not implemented");
        }
    }
}

File: app/src/main/java/me/devsaki/hentoid/fragments/pin/PinDialogFragment.kt
package me.devsaki.hentoid.fragments.pin

import android.content.Context
import android.os.Build
import android.os.Bundle
import android.os.CombinedVibration
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.StringRes
import androidx.fragment.app.DialogFragment
import me.devsaki.hentoid.R
import me.devsaki.hentoid.databinding.FragmentPinDialogBinding
import me.devsaki.hentoid.util.ThemeHelper
import java.security.InvalidParameterException

abstract class PinDialogFragment : DialogFragment() {
    private val pinValue = StringBuilder(4)
    private var binding: FragmentPinDialogBinding? = null

    protected abstract fun onPinAccept(pin: String)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ThemeHelper.setStyle(
            requireActivity(), this, STYLE_NORMAL, R.style.Theme_Light_PinEntryDialog
        )
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View? {
        binding = FragmentPinDialogBinding.inflate(inflater, container, false)
        binding?.apply {
            button0.setOnClickListener { onKeyClick("0") }
            button1.setOnClickListener { onKeyClick("1") }
            button2.setOnClickListener { onKeyClick("2") }
            button3.setOnClickListener { onKeyClick("3") }
            button4.setOnClickListener { onKeyClick("4") }
            button5.setOnClickListener { onKeyClick("5") }
            button6.setOnClickListener { onKeyClick("6") }
            button7.setOnClickListener { onKeyClick("7") }
            button8.setOnClickListener { onKeyClick("8") }
            button9.setOnClickListener { onKeyClick("9") }
            buttonBackspace.setOnClickListener { onBackspaceClick() }
        }
        return binding!!.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    override fun onStop() {
        super.onStop()
        dialog!!.cancel()
    }

    fun clearPin() {
        pinValue.setLength(0)
        binding?.apply {
            imagePlaceholder1.visibility = View.INVISIBLE
            imagePlaceholder2.visibility = View.INVISIBLE
            imagePlaceholder3.visibility = View.INVISIBLE
            imagePlaceholder4.visibility = View.INVISIBLE
        }
    }

    fun setHeaderText(@StringRes resId: Int) {
        binding?.textHeader?.setText(resId)
    }

    @Suppress("DEPRECATION")
    fun vibrate() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) {
            val vibrator = requireContext().getSystemService(Context.VIBRATOR_SERVICE) as Vibrator?
            vibrator?.vibrate(300)
        } else {
            (requireContext().getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager).apply {
                if (defaultVibrator.areAllPrimitivesSupported(VibrationEffect.Composition.PRIMITIVE_CLICK)) {
                    vibrate(
                        CombinedVibration.createParallel(
                            VibrationEffect.startComposition()
                                .addPrimitive(VibrationEffect.Composition.PRIMITIVE_CLICK)
                                .compose()
                        )
                    )
                }
            }
        }
    }

    private fun onKeyClick(s: String) {
        if (pinValue.length == 4) return
        pinValue.append(s)
        binding?.apply {
            when (pinValue.length) {
                1 -> imagePlaceholder1.visibility = View.VISIBLE
                2 -> imagePlaceholder2.visibility = View.VISIBLE
                3 -> imagePlaceholder3.visibility = View.VISIBLE
                4 -> {
                    imagePlaceholder4.visibility = View.VISIBLE
                    onPinAccept(pinValue.toString())
                }

                else -> throw InvalidParameterException("Not implemented")
            }
        }
    }

    private fun onBackspaceClick() {
        if (pinValue.isEmpty()) return
        pinValue.setLength(pinValue.length - 1)
        binding?.apply {
            when (pinValue.length) {
                0 -> imagePlaceholder1.visibility = View.INVISIBLE
                1 -> imagePlaceholder2.visibility = View.INVISIBLE
                2 -> imagePlaceholder3.visibility = View.INVISIBLE
                3 -> imagePlaceholder4.visibility = View.INVISIBLE
                else -> throw InvalidParameterException("Not implemented")
            }
        }
    }
}